### PR TITLE
fix: naming for services & service entrypoint

### PR DESCRIPTION
**Pull request by nielstenboom**
Hi me again 😄 

We ran into two bugs using the k8s hooks with a service, the bugs were as follows:
- Kubernetes would not accept the generated service name if you provide a non-official docker image (e.g localstack/localstack), Kubernetes then complains about a `/` being in the container name.
- The code would send `entrypoint: [undefined]` as the service spec, which Kubernetes would transform into `[""]` and that causes an error. This was not caught by the test because the actual args passed actually contained `entrypoint: null`. I've added that to the example.

Please let me know what you think!

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -1,12 +1,12 @@
 {
   "name": "hooks",
-  "version": "0.1.3",
+  "version": "0.2.1",
```
Comment (nikola-jokic) [inline package-lock.json:3]: Can you revert this back. This should be a part of the release and it should be in a separate PR :relaxed:

File: package-lock.json [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
   "lockfileVersion": 2,
   "requires": true,
   "packages": {
     "": {
       "name": "hooks",
-      "version": "0.1.3",
+      "version": "0.2.1",
```
Comment (nikola-jokic) [inline package-lock.json:9]: Same as above :point_up:

File: package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "hooks",
-  "version": "0.2.0",
+  "version": "0.2.1",
```
Comment (nikola-jokic) [inline package.json:3]: Can you please revert this back. The new release will update the version

File: packages/k8s/src/hooks/prepare-job.ts [hunk 4]
```diff
@@ -166,14 +167,17 @@ function createPodSpec(
...
     ports: containerPorts(container)
   } as k8s.V1Container
   if (container.workingDirectory) {
     podContainer.workingDir = container.workingDirectory
   }
 
+  if (container.entryPoint) {
+    podContainer.command = [container.entryPoint]
+  }
+
```
Comment (nikola-jokic) [inline packages/k8s/src/hooks/prepare-job.ts:180]: We should avoid setting up args from container and default to the same check like for the entrypoint:
```js
if (container.entryPointArgs?.length) {
    podContainer.args = container.entryPointArgs
}
```

**Review by nikola-jokic — CHANGES_REQUESTED** [review]
I left a few comments but otherwise looks great! Thanks for the contribution!

File: packages/k8s/src/hooks/prepare-job.ts (hunk 4 above)
Comment (nielstenboom) [reply packages/k8s/src/hooks/prepare-job.ts:179]: Good point!

**Comment by nielstenboom** [comment]
Thanks for the review @nikola-jokic ! I've implemented your suggestions 👍

File: packages/k8s/src/k8s/utils.ts [hunk 5]
```diff
@@ -137,6 +137,11 @@ exec ${environmentPrefix} ${entryPoint} ${
   }
 }
 
+export function generateContainerName(image: string): string {
+  const name = image.split('/').pop()
+  return name?.split(':').at(0) || 'service'
```
Comment (nikola-jokic) [inline packages/k8s/src/k8s/utils.ts:142]: I might be wrong on this one so please correct me if I am :relaxed: 
I'd suggest throwing an exception here.
We can return 'service' only if split(':') returns first element as faulty one. That can happen in case we have an empty image, or we have some bad input like: `something/:latest`. If we return 'service' here, then the pod will fail since the image is not created well, but we can avoid the API call entirely just by throwing exception here. We know that if there is no part we can name the container, the image is faulty for sure

File: packages/k8s/src/hooks/prepare-job.ts [hunk 6]
```diff
@@ -166,14 +167,20 @@ function createPodSpec(
   const podContainer = {
     name,
     image: container.image,
-    command: [container.entryPoint],
-    args: container.entryPointArgs,
     ports: containerPorts(container)
```
Comment (claveaur) [inline packages/k8s/src/hooks/prepare-job.ts:170]: What you guys think about adding same sort of check for ports like 
`if (container.portMappings) podContainer.ports = containerPorts(container)`
By default, I don't need to expose ports for containers in same pod, and currently if I don't specify a port it fails.

Comment (nikola-jokic) [reply packages/k8s/src/hooks/prepare-job.ts:170]: Hey @claveaur,

Can you please submit an issue with the repro? It should be part of a different PR. :relaxed:

File: packages/k8s/src/k8s/utils.ts (hunk 5 above)
Comment (nielstenboom) [reply packages/k8s/src/k8s/utils.ts]: Yep good point, agreed. Changed it and added some tests for it as well!

File: packages/k8s/src/k8s/utils.ts [hunk 7]
```diff
@@ -137,6 +137,17 @@ exec ${environmentPrefix} ${entryPoint} ${
   }
 }
 
+export function generateContainerName(image: string): string {
+  const nameWithTag = image.split('/').pop()
+  const name = nameWithTag?.split(':').at(0)
+
+  if (name === undefined || name === '') {
```
Comment (nikola-jokic) [inline packages/k8s/src/k8s/utils.ts:144]:
```suggestion
  if (!name) {
```
Sorry for being a stickler :relaxed: This is the last one

Comment (nielstenboom) [reply packages/k8s/src/k8s/utils.ts]: No worries 😄 

Done 👍 

Not sure if you can; but feel free to push changes to this branch if you happen to spot anything else!

**Review by nikola-jokic — APPROVED** [review]
LGTM

**Review by nikola-jokic — CHANGES_REQUESTED** [review]
Can you please fix the linter issue

**Comment by nielstenboom** [comment]
> Can you please fix the linter issue

Should be fixed now after merging main into this branch 👍 

It was caused by an outdated package-lock.json

**Review by nikola-jokic — APPROVED** [review]

File: packages/k8s/tests/k8s-utils-test.ts [hunk 8]
```diff
@@ -221,4 +225,32 @@ describe('k8s utils', () => {
...
+    it('should return the container name from image string', () => {
+      expect(
+        generateContainerName('public.ecr.aws/localstack/localstack')
+      ).toEqual('localstack')
+      expect(
+        generateContainerName(
+          'public.ecr.aws/url/with/multiple/slashes/postgres:latest'
+        )
+      ).toEqual('postgres')
+      expect(generateContainerName('postgres')).toEqual('postgres')
```
Comment (fhammerl) [inline packages/k8s/tests/k8s-utils-test.ts:239]: NIT: Could we use something like a `test.each` here instead of all `expect` in a single unit test?

File: packages/k8s/src/k8s/utils.ts [hunk 9]
```diff
@@ -137,6 +137,17 @@ exec ${environmentPrefix} ${entryPoint} ${
   }
 }
 
+export function generateContainerName(image: string): string {
+  const nameWithTag = image.split('/').pop()
+  const name = nameWithTag?.split(':').at(0)
+
+  if (!name) {
+    throw new Error('Image definition is invalid')
```
Comment (fhammerl) [inline packages/k8s/src/k8s/utils.ts:145]: NIT: log the invalid string

Comment (fhammerl) [inline packages/k8s/src/k8s/utils.ts:145]:
```suggestion
    throw new Error(`Image definition '${image}' is invalid`)
```

File: packages/k8s/tests/k8s-utils-test.ts (hunk 8 above)
Comment (nielstenboom) [reply packages/k8s/tests/k8s-utils-test.ts:239]: We can of course, but I'd like to counter argue that the tests I've written here are consistent with the rest of the codebase (I couldn't find another `test.each` and [tests a few lines above](https://github.com/actions/runner-container-hooks/blob/bab39cc413ef6751d58f6100694e8737cf45666e/packages/k8s/tests/k8s-utils-test.ts#L205-L216) also have all expect in the unit tests  😉 )

**Review by fhammerl — APPROVED** [review]
LGTM!

File: packages/k8s/tests/k8s-utils-test.ts (hunk 8 above)
Comment (fhammerl) [reply packages/k8s/tests/k8s-utils-test.ts:239]: We'll look at the tests in a later PR :)
